File: protoparser/src/test/java/com/squareup/protoparser/EnumElementTest.java
package com.squareup.protoparser;

import java.util.Arrays;
import java.util.Collections;
import org.junit.Test;

import static com.squareup.protoparser.OptionElement.Kind.BOOLEAN;
import static com.squareup.protoparser.OptionElement.Kind.STRING;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.Assert.fail;

public class EnumElementTest {
  @Test public void nameRequired() {
    try {
      EnumElement.builder().qualifiedName("Test").build();
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("name == null");
    }
  }

  @Test public void nameSetsQualifiedName() {
    EnumElement test = EnumElement.builder().name("Test").build();
    assertThat(test.name()).isEqualTo("Test");
    assertThat(test.qualifiedName()).isEqualTo("Test");
  }

  @Test public void nullBuilderValuesThrow() {
    try {
      EnumElement.builder().name(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("name == null");
    }
    try {
      EnumElement.builder().qualifiedName(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("qualifiedName == null");
    }
    try {
      EnumElement.builder().documentation(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("documentation == null");
    }
    try {
      EnumElement.builder().addConstant(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("constant == null");
    }
    try {
      EnumElement.builder().addConstants(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("constants == null");
    }
    try {
      EnumElement.builder().addConstants(Collections.<EnumConstantElement>singleton(null));
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("constant == null");
    }
    try {
      EnumElement.builder().addOption(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("option == null");
    }
    try {
      EnumElement.builder().addOptions(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("options == null");
    }
    try {
      EnumElement.builder().addOptions(Collections.<OptionElement>singleton(null));
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("option == null");
    }
  }

  @Test public void emptyToSchema() {
    EnumElement element = EnumElement.builder().name("Enum").build();
    String expected = "enum Enum {}\n";
    assertThat(element.toSchema()).isEqualTo(expected);
  }

  @Test public void simpleToSchema() {
    EnumElement element = EnumElement.builder()
        .name("Enum")
        .addConstant(EnumConstantElement.builder().name("ONE").tag(1).build())
        .addConstant(EnumConstantElement.builder().name("TWO").tag(2).build())
        .addConstant(EnumConstantElement.builder().name("SIX").tag(6).build())
        .build();
    String expected = ""
        + "enum Enum {\n"
        + "  ONE = 1;\n"
        + "  TWO = 2;\n"
        + "  SIX = 6;\n"
        + "}\n";
    assertThat(element.toSchema()).isEqualTo(expected);
  }

  @Test public void addMultipleConstants() {
    EnumConstantElement one = EnumConstantElement.builder().name("ONE").tag(1).build();
    EnumConstantElement two = EnumConstantElement.builder().name("TWO").tag(2).build();
    EnumConstantElement six = EnumConstantElement.builder().name("SIX").tag(6).build();
    EnumElement element = EnumElement.builder()
        .name("Enum")
        .addConstants(Arrays.asList(one, two, six))
        .build();
    assertThat(element.constants()).hasSize(3);
  }

  @Test public void simpleWithOptionsToSchema() {
    EnumElement element = EnumElement.builder()
        .name("Enum")
        .addOption(OptionElement.create("kit", STRING, "kat"))
        .addConstant(EnumConstantElement.builder().name("ONE").tag(1).build())
        .addConstant(EnumConstantElement.builder().name("TWO").tag(2).build())
        .addConstant(EnumConstantElement.builder().name("SIX").tag(6).build())
        .build();
    String expected = ""
        + "enum Enum {\n"
        + "  option kit = \"kat\";\n"
        + "\n"
        + "  ONE = 1;\n"
        + "  TWO = 2;\n"
        + "  SIX = 6;\n"
        + "}\n";
    assertThat(element.toSchema()).isEqualTo(expected);
  }

  @Test public void addMultipleOptions() {
    OptionElement kitKat = OptionElement.create("kit", STRING, "kat");
    OptionElement fooBar = OptionElement.create("foo", STRING, "bar");
    EnumElement element = EnumElement.builder()
        .name("Enum")
        .addOptions(Arrays.asList(kitKat, fooBar))
        .addConstant(EnumConstantElement.builder().name("ONE").tag(1).build())
        .build();
    assertThat(element.options()).hasSize(2);
  }

  @Test public void simpleWithDocumentationToSchema() {
    EnumElement element = EnumElement.builder()
        .name("Enum")
        .documentation("Hello")
        .addConstant(EnumConstantElement.builder().name("ONE").tag(1).build())
        .addConstant(EnumConstantElement.builder().name("TWO").tag(2).build())
        .addConstant(EnumConstantElement.builder().name("SIX").tag(6).build())
        .build();
    String expected = ""
        + "// Hello\n"
        + "enum Enum {\n"
        + "  ONE = 1;\n"
        + "  TWO = 2;\n"
        + "  SIX = 6;\n"
        + "}\n";
    assertThat(element.toSchema()).isEqualTo(expected);
  }

  @Test public void fieldToSchema() {
    EnumConstantElement value = EnumConstantElement.builder().name("NAME").tag(1).build();
    String expected = "NAME = 1;\n";
    assertThat(value.toSchema()).isEqualTo(expected);
  }

  @Test public void fieldWithDocumentationToSchema() {
    EnumConstantElement value = EnumConstantElement.builder()
        .name("NAME")
        .tag(1)
        .documentation("Hello")
        .build();
    String expected = ""
        + "// Hello\n"
        + "NAME = 1;\n";
    assertThat(value.toSchema()).isEqualTo(expected);
  }

  @Test public void fieldWithOptionsToSchema() {
    EnumConstantElement value = EnumConstantElement.builder()
        .name("NAME")
        .tag(1)
        .addOption(OptionElement.create("kit", STRING, "kat", true))
        .addOption(OptionElement.create("tit", STRING, "tat"))
        .build();
    String expected = "NAME = 1 [\n"
        + "  (kit) = \"kat\",\n"
        + "  tit = \"tat\"\n"
        + "];\n";
    assertThat(value.toSchema()).isEqualTo(expected);
  }

  @Test public void duplicateValueTagThrows() {
    try {
      EnumElement.builder()
          .name("Enum1")
          .qualifiedName("example.Enum")
          .addConstant(EnumConstantElement.builder().name("VALUE1").tag(1).build())
          .addConstant(EnumConstantElement.builder().name("VALUE2").tag(1).build())
          .build();
      fail();
    } catch (IllegalStateException e) {
      assertThat(e).hasMessage("Duplicate tag 1 in example.Enum");
    }
  }

  @Test public void duplicateValueTagWithAllowAlias() {
    EnumElement element = EnumElement.builder()
        .name("Enum1")
        .qualifiedName("example.Enum")
        .addOption(OptionElement.create("allow_alias", BOOLEAN, "true"))
        .addConstant(EnumConstantElement.builder().name("VALUE1").tag(1).build())
        .addConstant(EnumConstantElement.builder().name("VALUE2").tag(1).build())
        .build();
    assertThat(element.constants()).hasSize(2);
  }
}

File: wire-schema/src/commonTest/kotlin/com/squareup/wire/schema/internal/parser/EnumElementTest.kt
/*
 * Copyright (C) 2014 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire.schema.internal.parser

import assertk.assertThat
import assertk.assertions.hasSize
import assertk.assertions.isEqualTo
import com.squareup.wire.schema.Location
import com.squareup.wire.schema.internal.MAX_TAG_VALUE
import com.squareup.wire.schema.internal.parser.OptionElement.Kind.STRING
import kotlin.test.Test

class EnumElementTest {
  internal var location = Location.get("file.proto")

  @Test
  fun emptyToSchema() {
    val element = EnumElement(
      location = location,
      name = "Enum",
    )
    val expected = "enum Enum {}\n"
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun simpleToSchema() {
    val element = EnumElement(
      location = location,
      name = "Enum",
      constants = listOf(
        EnumConstantElement(location = location, name = "ONE", tag = 1),
        EnumConstantElement(location = location, name = "TWO", tag = 2),
        EnumConstantElement(location = location, name = "SIX", tag = 6),
      ),
    )
    val expected = """
        |enum Enum {
        |  ONE = 1;
        |  TWO = 2;
        |  SIX = 6;
        |}
        |
    """.trimMargin()
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun addMultipleConstants() {
    val one = EnumConstantElement(location = location, name = "ONE", tag = 1)
    val two = EnumConstantElement(location = location, name = "TWO", tag = 2)
    val six = EnumConstantElement(location = location, name = "SIX", tag = 6)
    val element = EnumElement(
      location = location,
      name = "Enum",
      constants = listOf(one, two, six),
    )
    assertThat(element.constants).hasSize(3)
  }

  @Test
  fun simpleWithOptionsToSchema() {
    val element = EnumElement(
      location = location,
      name = "Enum",
      options = listOf(OptionElement.create("kit", STRING, "kat")),
      constants = listOf(
        EnumConstantElement(location = location, name = "ONE", tag = 1),
        EnumConstantElement(location = location, name = "TWO", tag = 2),
        EnumConstantElement(location = location, name = "SIX", tag = 6),
      ),
    )
    val expected = """
        |enum Enum {
        |  option kit = "kat";
        |  ONE = 1;
        |  TWO = 2;
        |  SIX = 6;
        |}
        |
    """.trimMargin()
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun addMultipleOptions() {
    val kitKat = OptionElement.create("kit", STRING, "kat")
    val fooBar = OptionElement.create("foo", STRING, "bar")
    val element = EnumElement(
      location = location,
      name = "Enum",
      options = listOf(kitKat, fooBar),
      constants = listOf(EnumConstantElement(location = location, name = "ONE", tag = 1)),
    )
    assertThat(element.options).hasSize(2)
  }

  @Test
  fun simpleWithDocumentationToSchema() {
    val element = EnumElement(
      location = location,
      name = "Enum",
      documentation = "Hello",
      constants =
      listOf(
        EnumConstantElement(location = location, name = "ONE", tag = 1),
        EnumConstantElement(location = location, name = "TWO", tag = 2),
        EnumConstantElement(location = location, name = "SIX", tag = 6),
      ),
    )
    val expected = """
        |// Hello
        |enum Enum {
        |  ONE = 1;
        |  TWO = 2;
        |  SIX = 6;
        |}
        |
    """.trimMargin()
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun simpleWithReservedToSchema() {
    val element = EnumElement(
      location = location,
      name = "Enum",
      reserveds = listOf(
        ReservedElement(location = location, values = listOf(10, 12..14, "FOO")),
        ReservedElement(location = location, values = listOf(10)),
        ReservedElement(location = location, values = listOf(25..MAX_TAG_VALUE)),
        ReservedElement(location = location, values = listOf("FOO")),
      ),
      constants = listOf(
        EnumConstantElement(location = location, name = "ONE", tag = 1),
        EnumConstantElement(location = location, name = "TWO", tag = 2),
        EnumConstantElement(location = location, name = "SIX", tag = 6),
      ),
    )
    val expected = """
        |enum Enum {
        |  reserved 10, 12 to 14, "FOO";
        |  reserved 10;
        |  reserved 25 to max;
        |  reserved "FOO";
        |  ONE = 1;
        |  TWO = 2;
        |  SIX = 6;
        |}
        |
    """.trimMargin()
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun fieldToSchema() {
    val value = EnumConstantElement(location = location, name = "NAME", tag = 1)
    val expected = "NAME = 1;\n"
    assertThat(value.toSchema()).isEqualTo(expected)
  }

  @Test
  fun fieldWithDocumentationToSchema() {
    val value = EnumConstantElement(
      location = location,
      name = "NAME",
      tag = 1,
      documentation = "Hello",
    )
    val expected = """
        |// Hello
        |NAME = 1;
        |
    """.trimMargin()
    assertThat(value.toSchema()).isEqualTo(expected)
  }

  @Test
  fun fieldWithOptionsToSchema() {
    val value = EnumConstantElement(
      location = location,
      name = "NAME",
      tag = 1,
      options = listOf(
        OptionElement.create("kit", STRING, "kat", true),
        OptionElement.create("tit", STRING, "tat"),
      ),
    )
    val expected = """
        |NAME = 1 [
        |  (kit) = "kat",
        |  tit = "tat"
        |];
        |
    """.trimMargin()
    assertThat(value.toSchema()).isEqualTo(expected)
  }
}
